Using @ConfigProperty with Google Cloud Function throws Exception when executing QuarkusTest

### Describe the bug

When using Google Cloud HTTP Function in combination with a `ConfigProperty`, an Exception is thrown when executing a QuarkusTest calling that Google Cloud HTTP Function. 
This used to work fine in Quarkus Platform version 3.31.4. 
This is not working anymore in Quarkus Platform version 3.32.2 and upwards, including the latest version 3.34.6.

The [sample code](https://github.com/tinohertlein/quarkus-issue) demonstrates this issue by extending the [Quarkus quickstart example](https://github.com/quarkusio/quarkus-quickstarts/tree/main/google-cloud-functions-quickstart) with

- a ConfigProperty
- a test that calls the Google Cloud HTTP Function

### Expected behavior

`QuarkusTest` runs without throwing an Exception.

### Actual behavior

`QuarkusTest` throws the following Exception:

```
Suppressed: java.lang.RuntimeException: Error injecting java.lang.String org.acme.quickstart.service.GreetingService.greeterName 
Caused by: java.lang.IllegalArgumentException: SRCFG00015: No configuration is available for this class loader
```

### How to Reproduce?

Clone the [sample project](https://github.com/tinohertlein/quarkus-issue) and execute the test in [src/test/java/HttpFunctionGreeterTest.java](src/test/java/HttpFunctionGreeterTest.java)

### Output of `uname -a` or `ver`

Darwin MacBook-Pro 25.1.0 Darwin Kernel Version 25.1.0: Mon Oct 20 19:32:41 PDT 2025; root:xnu-12377.41.6~2/RELEASE_ARM64_T6000 arm64

### Output of `java -version`

openjdk 25.0.2 2026-01-20 LTS

### Quarkus version or git rev

Quarkus Platform version 3.32.2 and upwards

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.11

### Additional information

This used to work fine in Quarkus Platform version 3.31.4. 
This is not working anymore in Quarkus Platform version 3.32.2 and upwards, including the latest version 3.34.6.

--- a/extensions/google-cloud-functions/deployment/src/main/java/io/quarkus/gcp/functions/deployment/GoogleCloudFunctionsProcessor.java
+++ b/extensions/google-cloud-functions/deployment/src/main/java/io/quarkus/gcp/functions/deployment/GoogleCloudFunctionsProcessor.java
@@ -27,6 +27,7 @@ import io.quarkus.deployment.annotations.Record;
 import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
 import io.quarkus.deployment.builditem.FeatureBuildItem;
 import io.quarkus.deployment.builditem.RunTimeConfigurationDefaultBuildItem;
+import io.quarkus.deployment.builditem.ShutdownContextBuildItem;
 import io.quarkus.gcp.functions.GoogleCloudFunctionInfo;
 import io.quarkus.gcp.functions.GoogleCloudFunctionRecorder;
 
@@ -95,12 +96,13 @@ public class GoogleCloudFunctionsProcessor {
 
     @BuildStep
     @Record(ExecutionTime.RUNTIME_INIT)
-    public void selectDelegate(List<CloudFunctionBuildItem> cloudFunctions, GoogleCloudFunctionRecorder recorder) {
+    public void selectDelegate(List<CloudFunctionBuildItem> cloudFunctions, GoogleCloudFunctionRecorder recorder,
+            ShutdownContextBuildItem shutdown) {
 
         List<GoogleCloudFunctionInfo> functionInfos = cloudFunctions.stream()
                 .map(CloudFunctionBuildItem::build)
                 .collect(Collectors.toList());
 
-        recorder.selectDelegate(functionInfos);
+        recorder.selectDelegate(functionInfos, shutdown);
     }
 }

--- a/extensions/google-cloud-functions/runtime/src/main/java/io/quarkus/gcp/functions/GoogleCloudFunctionRecorder.java
+++ b/extensions/google-cloud-functions/runtime/src/main/java/io/quarkus/gcp/functions/GoogleCloudFunctionRecorder.java
@@ -5,6 +5,7 @@ import java.util.List;
 import java.util.Map;
 
 import io.quarkus.runtime.RuntimeValue;
+import io.quarkus.runtime.ShutdownContext;
 import io.quarkus.runtime.annotations.Recorder;
 
 @Recorder
@@ -15,7 +16,8 @@ public class GoogleCloudFunctionRecorder {
         this.runtimeConfig = runtimeConfig;
     }
 
-    public void selectDelegate(List<GoogleCloudFunctionInfo> cloudFunctions) {
+    public void selectDelegate(List<GoogleCloudFunctionInfo> cloudFunctions,
+            ShutdownContext shutdownContext) {
         Map<GoogleCloudFunctionInfo.FunctionType, String> delegates = new HashMap<>();
         // if a function name is defined, check that it exists
         if (runtimeConfig.getValue().function().isPresent()) {
@@ -44,5 +46,14 @@ public class GoogleCloudFunctionRecorder {
         QuarkusBackgroundFunction.setDelegates(delegates.get(GoogleCloudFunctionInfo.FunctionType.BACKGROUND),
                 delegates.get(GoogleCloudFunctionInfo.FunctionType.RAW_BACKGROUND));
         QuarkusCloudEventsFunction.setDelegate(delegates.get(GoogleCloudFunctionInfo.FunctionType.CLOUD_EVENT));
+
+        shutdownContext.addShutdownTask(new Runnable() {
+            @Override
+            public void run() {
+                QuarkusHttpFunction.clearState();
+                QuarkusBackgroundFunction.clearState();
+                QuarkusCloudEventsFunction.clearState();
+            }
+        });
     }
 }

--- a/extensions/google-cloud-functions/runtime/src/main/java/io/quarkus/gcp/functions/QuarkusBackgroundFunction.java
+++ b/extensions/google-cloud-functions/runtime/src/main/java/io/quarkus/gcp/functions/QuarkusBackgroundFunction.java
@@ -14,10 +14,11 @@ import com.google.gson.Gson;
 import io.quarkus.arc.Arc;
 import io.quarkus.runtime.Application;
 
+@SuppressWarnings({ "rawtypes", "unchecked" })
 public final class QuarkusBackgroundFunction implements RawBackgroundFunction {
 
-    protected static final String deploymentStatus;
-    protected static boolean started = false;
+    private static final String deploymentStatus;
+    private static boolean started = false;
 
     private static volatile BackgroundFunction delegate;
     private static volatile Class<?> parameterType;

--- a/extensions/google-cloud-functions/runtime/src/main/java/io/quarkus/gcp/functions/QuarkusBackgroundFunction.java
+++ b/extensions/google-cloud-functions/runtime/src/main/java/io/quarkus/gcp/functions/QuarkusBackgroundFunction.java
@@ -22,6 +22,7 @@ public final class QuarkusBackgroundFunction implements RawBackgroundFunction {
     private static volatile BackgroundFunction delegate;
     private static volatile Class<?> parameterType;
     private static volatile RawBackgroundFunction rawDelegate;
+    private static volatile ClassLoader delegateClassLoader;
 
     static {
         StringWriter error = new StringWriter();
@@ -53,9 +54,10 @@ public final class QuarkusBackgroundFunction implements RawBackgroundFunction {
     }
 
     static void setDelegates(String selectedDelegate, String selectedRawDelegate) {
+        delegateClassLoader = Thread.currentThread().getContextClassLoader();
         if (selectedDelegate != null) {
             try {
-                Class<?> clazz = Class.forName(selectedDelegate, false, Thread.currentThread().getContextClassLoader());
+                Class<?> clazz = Class.forName(selectedDelegate, false, delegateClassLoader);
                 for (Method method : clazz.getDeclaredMethods()) {
                     if (method.getName().equals("accept")) {
                         // the first parameter of the accept method is the event, we need to register it's type to
@@ -74,7 +76,7 @@ public final class QuarkusBackgroundFunction implements RawBackgroundFunction {
 
         if (selectedRawDelegate != null) {
             try {
-                Class<?> clazz = Class.forName(selectedRawDelegate, false, Thread.currentThread().getContextClassLoader());
+                Class<?> clazz = Class.forName(selectedRawDelegate, false, delegateClassLoader);
                 rawDelegate = (RawBackgroundFunction) Arc.container().instance(clazz).get();
             } catch (ClassNotFoundException e) {
                 throw new RuntimeException(e);
@@ -94,22 +96,23 @@ public final class QuarkusBackgroundFunction implements RawBackgroundFunction {
                     "(or there is multiple one and none selected inside your application.properties)");
         }
 
-        if (rawDelegate != null) {
-            rawDelegate.accept(event, context);
-        } else {
-            // here we emulate a BackgroundFunction from a RawBackgroundFunction
-            // so we need to un-marshall the JSON into the delegate parameter type
-            // this is done via Gson as it's the library that GCF uses internally for the same purpose
-            // see NewBackgroundFunctionExecutor.TypedFunctionExecutor<T>
-            // from the functions-framework-java https://github.com/GoogleCloudPlatform/functions-framework-java
-            Gson gson = new Gson();
-            try {
-                Object eventObj = gson.fromJson(event, parameterType);
-                delegate.accept(eventObj, context);
-            } catch (JsonParseException e) {
-                throw new RuntimeException("Could not parse received event payload into type "
-                        + parameterType.getCanonicalName(), e);
+        ClassLoader currentCl = Thread.currentThread().getContextClassLoader();
+        try {
+            Thread.currentThread().setContextClassLoader(delegateClassLoader);
+            if (rawDelegate != null) {
+                rawDelegate.accept(event, context);
+            } else {
+                Gson gson = new Gson();
+                try {
+                    Object eventObj = gson.fromJson(event, parameterType);
+                    delegate.accept(eventObj, context);
+                } catch (JsonParseException e) {
+                    throw new RuntimeException("Could not parse received event payload into type "
+                            + parameterType.getCanonicalName(), e);
+                }
             }
+        } finally {
+            Thread.currentThread().setContextClassLoader(currentCl);
         }
     }
 }

--- a/extensions/google-cloud-functions/runtime/src/main/java/io/quarkus/gcp/functions/QuarkusBackgroundFunction.java
+++ b/extensions/google-cloud-functions/runtime/src/main/java/io/quarkus/gcp/functions/QuarkusBackgroundFunction.java
@@ -85,6 +85,13 @@ public final class QuarkusBackgroundFunction implements RawBackgroundFunction {
         }
     }
 
+    public static void clearState() {
+        delegate = null;
+        parameterType = null;
+        rawDelegate = null;
+        delegateClassLoader = null;
+    }
+
     @Override
     public void accept(String event, Context context) throws Exception {
         if (!started) {

--- a/extensions/google-cloud-functions/runtime/src/main/java/io/quarkus/gcp/functions/QuarkusCloudEventsFunction.java
+++ b/extensions/google-cloud-functions/runtime/src/main/java/io/quarkus/gcp/functions/QuarkusCloudEventsFunction.java
@@ -12,8 +12,8 @@ import io.quarkus.runtime.Application;
 
 public final class QuarkusCloudEventsFunction implements CloudEventsFunction {
 
-    protected static final String deploymentStatus;
-    protected static boolean started = false;
+    private static final String deploymentStatus;
+    private static boolean started = false;
 
     private static volatile CloudEventsFunction delegate;
     private static volatile ClassLoader delegateClassLoader;

--- a/extensions/google-cloud-functions/runtime/src/main/java/io/quarkus/gcp/functions/QuarkusCloudEventsFunction.java
+++ b/extensions/google-cloud-functions/runtime/src/main/java/io/quarkus/gcp/functions/QuarkusCloudEventsFunction.java
@@ -16,6 +16,7 @@ public final class QuarkusCloudEventsFunction implements CloudEventsFunction {
     protected static boolean started = false;
 
     private static volatile CloudEventsFunction delegate;
+    private static volatile ClassLoader delegateClassLoader;
 
     static {
         StringWriter error = new StringWriter();
@@ -48,8 +49,9 @@ public final class QuarkusCloudEventsFunction implements CloudEventsFunction {
 
     static void setDelegate(String selectedDelegate) {
         if (selectedDelegate != null) {
+            delegateClassLoader = Thread.currentThread().getContextClassLoader();
             try {
-                Class<?> clazz = Class.forName(selectedDelegate, false, Thread.currentThread().getContextClassLoader());
+                Class<?> clazz = Class.forName(selectedDelegate, false, delegateClassLoader);
                 delegate = (CloudEventsFunction) Arc.container().instance(clazz).get();
             } catch (ClassNotFoundException e) {
                 throw new RuntimeException(e);
@@ -69,6 +71,12 @@ public final class QuarkusCloudEventsFunction implements CloudEventsFunction {
                     "(or there is multiple one and none selected inside your application.properties)");
         }
 
-        delegate.accept(cloudEvent);
+        ClassLoader currentCl = Thread.currentThread().getContextClassLoader();
+        try {
+            Thread.currentThread().setContextClassLoader(delegateClassLoader);
+            delegate.accept(cloudEvent);
+        } finally {
+            Thread.currentThread().setContextClassLoader(currentCl);
+        }
     }
 }

--- a/extensions/google-cloud-functions/runtime/src/main/java/io/quarkus/gcp/functions/QuarkusCloudEventsFunction.java
+++ b/extensions/google-cloud-functions/runtime/src/main/java/io/quarkus/gcp/functions/QuarkusCloudEventsFunction.java
@@ -59,6 +59,11 @@ public final class QuarkusCloudEventsFunction implements CloudEventsFunction {
         }
     }
 
+    public static void clearState() {
+        delegate = null;
+        delegateClassLoader = null;
+    }
+
     @Override
     public void accept(CloudEvent cloudEvent) throws Exception {
         if (!started) {

--- a/extensions/google-cloud-functions/runtime/src/main/java/io/quarkus/gcp/functions/QuarkusHttpFunction.java
+++ b/extensions/google-cloud-functions/runtime/src/main/java/io/quarkus/gcp/functions/QuarkusHttpFunction.java
@@ -17,6 +17,7 @@ public final class QuarkusHttpFunction implements HttpFunction {
     protected static boolean started = false;
 
     private static volatile HttpFunction delegate;
+    private static volatile ClassLoader delegateClassLoader;
 
     static {
         StringWriter error = new StringWriter();
@@ -49,8 +50,9 @@ public final class QuarkusHttpFunction implements HttpFunction {
 
     static void setDelegate(String selectedDelegate) {
         if (selectedDelegate != null) {
+            delegateClassLoader = Thread.currentThread().getContextClassLoader();
             try {
-                Class<?> clazz = Class.forName(selectedDelegate, false, Thread.currentThread().getContextClassLoader());
+                Class<?> clazz = Class.forName(selectedDelegate, false, delegateClassLoader);
                 delegate = (HttpFunction) Arc.container().instance(clazz).get();
             } catch (ClassNotFoundException e) {
                 throw new RuntimeException(e);
@@ -70,6 +72,12 @@ public final class QuarkusHttpFunction implements HttpFunction {
                     "(or there is multiple one and none selected inside your application.properties)");
         }
 
-        delegate.service(httpRequest, httpResponse);
+        ClassLoader currentCl = Thread.currentThread().getContextClassLoader();
+        try {
+            Thread.currentThread().setContextClassLoader(delegateClassLoader);
+            delegate.service(httpRequest, httpResponse);
+        } finally {
+            Thread.currentThread().setContextClassLoader(currentCl);
+        }
     }
 }

--- a/extensions/google-cloud-functions/runtime/src/main/java/io/quarkus/gcp/functions/QuarkusHttpFunction.java
+++ b/extensions/google-cloud-functions/runtime/src/main/java/io/quarkus/gcp/functions/QuarkusHttpFunction.java
@@ -13,8 +13,8 @@ import io.quarkus.runtime.Application;
 
 public final class QuarkusHttpFunction implements HttpFunction {
 
-    protected static final String deploymentStatus;
-    protected static boolean started = false;
+    private static final String deploymentStatus;
+    private static boolean started = false;
 
     private static volatile HttpFunction delegate;
     private static volatile ClassLoader delegateClassLoader;

--- a/extensions/google-cloud-functions/runtime/src/main/java/io/quarkus/gcp/functions/QuarkusHttpFunction.java
+++ b/extensions/google-cloud-functions/runtime/src/main/java/io/quarkus/gcp/functions/QuarkusHttpFunction.java
@@ -60,6 +60,11 @@ public final class QuarkusHttpFunction implements HttpFunction {
         }
     }
 
+    static void clearState() {
+        delegate = null;
+        delegateClassLoader = null;
+    }
+
     @Override
     public void service(HttpRequest httpRequest, HttpResponse httpResponse) throws Exception {
         if (!started) {

--- a/integration-tests/google-cloud-functions/src/main/java/io/quarkus/gcp/function/test/service/GreetingService.java
+++ b/integration-tests/google-cloud-functions/src/main/java/io/quarkus/gcp/function/test/service/GreetingService.java
@@ -2,9 +2,15 @@ package io.quarkus.gcp.function.test.service;
 
 import jakarta.enterprise.context.ApplicationScoped;
 
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+
 @ApplicationScoped
 public class GreetingService {
+
+    @ConfigProperty(name = "greeting.name", defaultValue = "World")
+    String name;
+
     public String hello() {
-        return "Hello World!";
+        return "Hello " + name + "!";
     }
 }

--- a/integration-tests/google-cloud-functions/src/main/resources/application.properties
+++ b/integration-tests/google-cloud-functions/src/main/resources/application.properties
@@ -4,3 +4,5 @@ quarkus.google-cloud-functions.function=httpTest
 %background.quarkus.google-cloud-functions.function=storageTest
 
 quarkus.http.test-port=8082
+
+greeting.name=Quarkus

--- a/integration-tests/google-cloud-functions/src/test/java/io/quarkus/gcp/function/test/HttpFunctionRandomPortTestCase.java
+++ b/integration-tests/google-cloud-functions/src/test/java/io/quarkus/gcp/function/test/HttpFunctionRandomPortTestCase.java
@@ -26,6 +26,6 @@ class HttpFunctionRandomPortTestCase {
                 .get()
                 .then()
                 .statusCode(200)
-                .body(is("Hello World!"));
+                .body(is("Hello Quarkus!"));
     }
 }

--- a/integration-tests/google-cloud-functions/src/test/java/io/quarkus/gcp/function/test/HttpFunctionTestCase.java
+++ b/integration-tests/google-cloud-functions/src/test/java/io/quarkus/gcp/function/test/HttpFunctionTestCase.java
@@ -19,7 +19,7 @@ class HttpFunctionTestCase {
                 .get()
                 .then()
                 .statusCode(200)
-                .body(is("Hello World!"));
+                .body(is("Hello Quarkus!"));
     }
 
     // we do twice the test to be sure we can call multiple time the same function
@@ -30,6 +30,6 @@ class HttpFunctionTestCase {
                 .get()
                 .then()
                 .statusCode(200)
-                .body(is("Hello World!"));
+                .body(is("Hello Quarkus!"));
     }
 }

